ruma-appservice-api/src/lib.rs

```diff
@@ -4,6 +4,69 @@
 
 #![warn(missing_debug_implementations, missing_docs)]
 
+use serde::{Deserialize, Serialize};
+
 pub mod event;
 pub mod query;
 pub mod thirdparty;
+
+/// A namespace defined by an application service.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Namespace {
+    /// A true or false value stating whether this application service has
+    /// exclusive access to events within this namespace.
+    pub exclusive: bool,
+
+    /// A regular expression defining which values this namespace includes.
+    pub regex: String,
+}
+
+/// Namespaces defined by an application service.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Namespaces {
+    /// Events which are sent from certain users.
+    pub users: Vec<Namespace>,
+
+    /// Events which are sent in rooms with certain room aliases.
+    pub aliases: Vec<Namespace>,
+
+    /// Events which are sent in rooms with certain room IDs.
+    pub rooms: Vec<Namespace>,
+}
+
+/// Information required in the registration yaml file that a homeserver needs.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Registration {
+    /// A unique, user - defined ID of the application service which will never change.
+    pub id: String,
+
+    /// The URL for the application service.
+    pub url: String,
+
+    /// A unique token for application services to use to authenticate requests to Homeservers.
+    pub as_token: String,
+
+    /// A unique token for Homeservers to use to authenticate requests to application services.
+    pub hs_token: String,
+
+    /// The localpart of the user associated with the application service.
+    pub sender_localpart: String,
+
+    /// A list of users, aliases and rooms namespaces that the application service controls.
+    pub namespaces: Namespaces,
+
+    /// Whether requests from masqueraded users are rate-limited. The sender is excluded.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rate_limited: Option<bool>,
+
+    /// The external protocols which the application service provides (e.g. IRC).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub protocols: Option<Vec<String>>,
+}
+
+impl Registration {
+    /// Create a `Registration` from the given yaml value.
+    pub fn from_yaml_str(s: &str) -> Result<Self, serde_yaml::Error> {
```

Comment:
I don't think this really adds anything, users can just use `serde_yaml::from_str` themselves. `serde_yaml` should only be a dev-dependency.

Reply:
Yeah, my thoughts as well, I had copied it from matrix-appservice-rs.